libnetwork/ipamutils/utils.go

```diff
@@ -71,52 +79,146 @@ func GetLocalScopeDefaultNetworks() []*net.IPNet {
 // ConfigGlobalScopeDefaultNetworks configures global default pool.
 // Ideally this will be called from SwarmKit as part of swarm init
 func ConfigGlobalScopeDefaultNetworks(defaultAddressPool []*NetworkToSplit) error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
 	if defaultAddressPool == nil {
 		defaultAddressPool = globalScopeDefaultNetworks
 	}
-	return configDefaultNetworks(defaultAddressPool, &PredefinedGlobalScopeDefaultNetworks)
+	var err error
+	PredefinedGlobalScopeDefaultNetworks, err = NewNetworkSplitter(defaultAddressPool)
+	return err
 }
 
 // ConfigLocalScopeDefaultNetworks configures local default pool.
 // Ideally this will be called during libnetwork init
 func ConfigLocalScopeDefaultNetworks(defaultAddressPool []*NetworkToSplit) error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
 	if defaultAddressPool == nil {
 		return nil
 	}
-	return configDefaultNetworks(defaultAddressPool, &PredefinedLocalScopeDefaultNetworks)
+	var err error
+	PredefinedLocalScopeDefaultNetworks, err = NewNetworkSplitter(defaultAddressPool)
+	return err
 }
 
-// splitNetworks takes a slice of networks, split them accordingly and returns them
-func splitNetworks(list []*NetworkToSplit) ([]*net.IPNet, error) {
-	localPools := make([]*net.IPNet, 0, len(list))
+// NetworkSplitter is an iterator producing subnets from its NetworkToSplit.
+// Its purpose is to avoid generating the whole list of subnets from NetworkToSplit and instead generate subnet one at
+// a time to avoid consuming way too much memory when NetworkToSplit contains large IPv6 ranges (eg. fd00::/8 split
+// into /96). See https://github.com/moby/moby/issues/40275.
+type NetworkSplitter struct {
```

Comment:
I would feel better if we had had a mutex in here or at least an explicit call-out that additional synchronization may be required and how it is used safely.